goinp/goinp.go

```diff
@@ -64,45 +65,72 @@ func AskForString(messageToPrint string) (string, error) {
 	return AskForStringFromReader(messageToPrint, os.Stdin)
 }
 
-func writeStdinClearable(text string) error {
+type stdin interface {
+	Fd() uintptr
+	Read([]byte) (int, error)
+}
+
+func writeStdinClearable(text string, in stdin) error {
 	// clearable stdin text is only possible if there is an available terminal input buffer
-	if terminal.IsTerminal(int(os.Stdin.Fd())) {
+	if terminal.IsTerminal(int(in.Fd())) {
 		for _, c := range []byte(text) {
-			if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, os.Stdin.Fd(), syscall.TIOCSTI, uintptr(unsafe.Pointer(&c))); errno != 0 {
+			if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, in.Fd(), syscall.TIOCSTI, uintptr(unsafe.Pointer(&c))); errno != 0 {
 				return fmt.Errorf("failed to write to stdin, err no: %d", errno)
 			}
 		}
 	}
 	return nil
 }
 
-func askOptionsWithReader(title string, defaultValue string, optional bool, reader *bufio.Reader, options ...string) (string, error) {
-	const customValueOptionText = "<custom value>"
+func readUntil(reader stdin, str string) (string, error) {
+	var strb []byte
+	for {
+		b := make([]byte, 1)
+		if _, err := reader.Read(b); err != nil {
+			return "", err
+		}
+		strb = append(strb, b[0])
+		if bytes.HasSuffix(strb, []byte(str)) {
+			return string(strb), nil
+		}
+	}
+}
 
-	if len(options) == 0 {
-		for {
-			if title != "" {
-				fmt.Print("Enter value for \"" + strings.TrimSuffix(title, ":") + "\": ")
+func askForInput(title string, defaultValue string, optional bool, reader stdin, writer io.Writer) (string, error) {
+	for {
+		if title != "" {
+			if _, err := fmt.Fprint(writer, "Enter value for \""+strings.TrimSuffix(title, ":")+"\": "); err != nil {
+				return "", err
 			}
+		}
 
-			if defaultValue != "" {
-				if err := writeStdinClearable(defaultValue); err != nil {
-					return "", err
-				}
+		if defaultValue != "" {
+			if err := writeStdinClearable(defaultValue, reader); err != nil {
+				return "", err
 			}
+		}
 
-			input, err := reader.ReadString('\n')
-			if err != nil {
+		input, err := readUntil(reader, "\n")
+		if err != nil {
+			return "", err
+		}
+
+		if !optional && strings.TrimSpace(input) == "" {
+			if _, err := fmt.Fprint(writer, colorstring.Red("value must be specified")+"\n"); err != nil {
 				return "", err
 			}
+			continue
+		}
 
-			if !optional && strings.TrimSpace(input) == "" {
-				fmt.Print(colorstring.Red("value must be specified") + "\n")
-				continue
-			}
+		return strings.TrimSpace(input), nil
+	}
+}
 
-			return strings.TrimSpace(input), nil
-		}
+func askOptions(title string, defaultValue string, optional bool, reader stdin, writer io.Writer, options ...string) (string, error) {
```

Comment:
why do we merge selecting from possible values and free inputs?
i would recommend having separate functions for these functionalities, to make the client side usage straightforward.